src/software/ai/hl/stp/tactic/receiver/receiver_fsm.h

```diff
@@ -0,0 +1,296 @@
+        const auto onetouch_s  = state<OneTouchShotState>;
+        const auto update_e    = event<Update>;
+
+        /**
+         * Checks if a one touch shot is possible
+         *
+         * @param event ReceiverFSM::Update event
+         * @return true if one-touch possible
+         */
+        const auto onetouch_possible = [](auto event) {
+            return findFeasibleShot(event.common.world, event.common.robot) !=
+                   std::nullopt;
+        };
+
+        /**
+         * If we have a shot on net, then update the receiver fsm
+         * to setup for a one-touch shot.
+         *
+         * NOTE: This must be used with the onetouch_possible guard,
+         * which checks for one-touch feasibility.
+         *
+         * @param event ReceiverFSM::Update event
+         */
+        const auto update_onetouch = [this](auto event) {
+            auto best_shot = findFeasibleShot(event.common.world, event.common.robot);
+            auto one_touch = getOneTimeShotPositionAndOrientation(
+                event.common.robot, event.common.world.ball(),
+                best_shot->getPointToShootAt());
+
+            if (best_shot)
+            {
+                if (event.control_params.pass)
+                {
+                    event.common.set_intent(std::make_unique<MoveIntent>(
+                        event.common.robot.id(), one_touch.getPointToShootAt(),
+                        one_touch.getOpenAngle(), 0, DribblerMode::OFF,
+                        BallCollisionType::ALLOW,
+                        AutoChipOrKick{AutoChipOrKickMode::AUTOKICK,
+                                       MAX_SPEED_FOR_ONE_TOUCH_SHOT},
+                        MaxAllowedSpeedMode::PHYSICAL_LIMIT, 0.0));
+                }
+            }
+        };
+
+        /**
+         * One touch shot is not possible, just receive ball as cleanly as possible.
+         *
+         * @param event ReceiverFSM::Update event
+         */
+        const auto update_receive = [this](auto event) {
+            if (event.control_params.pass)
+            {
+                event.common.set_intent(std::make_unique<MoveIntent>(
+                    event.common.robot.id(), event.control_params.pass->receiverPoint(),
+                    event.control_params.pass->receiverOrientation(), 0,
+                    DribblerMode::OFF, BallCollisionType::ALLOW,
+                    AutoChipOrKick{AutoChipOrKickMode::OFF, 0},
+                    MaxAllowedSpeedMode::PHYSICAL_LIMIT, 0.0));
+            }
+        };
+
+        const auto adjust_receive = [this](auto event) {
+            auto ball      = event.common.world.ball();
+            auto robot_pos = event.common.robot.position();
+
+            if ((ball.position() - robot_pos).length() >
+                DIST_TO_FRONT_OF_ROBOT_METERS + 2 * BALL_MAX_RADIUS_METERS)
+            {
+                Point ball_receive_pos = ball.position();
+
+                if (ball.velocity().length() != 0)
```

Comment:
should there be a threshold?

Reply:
probably doesn't matter too much if the ball velocity is like 0.001 m/s but we don't really need a threshold because we want the receiver to keep adjusting until the ball comes to a complete stop

Reply:
in real vision, the ball velocity will never be 0...

Reply:
this needs to filter out cases where `Line(ball.position(), ball.position() + ball.velocity())`fails, i.e. when `ball.position()==ball.position() + ball.velocity()`, which it doesn't currently

Reply:
bumping